crates/evm/evm/src/executors/invariant/mod.rs

```diff
@@ -297,6 +300,107 @@ pub struct InvariantExecutor<'a> {
     project_contracts: &'a ContractsByArtifact,
     /// Filters contracts to be fuzzed through their artifact identifiers.
     artifact_filters: ArtifactFilters,
+    /// History of binned hitcount of edges seen during fuzzing.
+    history_map: Vec<u8>,
+}
+const COVERAGE_MAP_SIZE: usize = 65536;
+
+pub struct Corpus {
+    pub corpus_dir: Option<PathBuf>,
+    pub corpus: Vec<Vec<BasicTxDetails>>,
+    pub tx_generator: BoxedStrategy<BasicTxDetails>,
+}
+
+impl Corpus {
+    pub fn new(corpus_dir: PathBuf, tx_generator: BoxedStrategy<BasicTxDetails>) -> Self {
+        // assert!(corpus_dir.is_dir());
+        let mut corpus = vec![];
+        if let Ok(entries) = std::fs::read_dir(&corpus_dir) {
+            for entry in entries {
+                let path = entry.expect("msg").path();
+                if path.is_file() && path.extension() == Some(std::ffi::OsStr::new("json")) {
+                    let tx_seq: Vec<BasicTxDetails> =
+                        foundry_common::fs::read_json_file(&path).expect("msg");
+                    corpus.push(tx_seq);
+                }
+            }
+        }
+        Self { corpus_dir: Some(corpus_dir), corpus, tx_generator }
+    }
+
+    // pub fn save(&self) -> Result<()> { TODO order or key by timestamp?
+    //     for (i, tx_seq) in self.corpus.iter().enumerate() {
+    //         let path = self.corpus_dir.join(format!("{}.json", i));
+    //         foundry_common::fs::write_json_file(&path, tx_seq)?;
+    //     }
+    //     Ok(())
+    // }
+
+    pub fn insert(&mut self, tx_seq: Vec<BasicTxDetails>) {
+        self.corpus.push(tx_seq);
+    }
+
+    pub fn new_sequence(&self, test_runnner: &mut TestRunner) -> Vec<BasicTxDetails> {
+        let rng = test_runnner.rng();
+
+        if self.corpus.len() > 1 {
+            let idx1 = rng.gen_range(0..self.corpus.len());
+            let idx2 = rng.gen_range(0..self.corpus.len());
+            let one = &self.corpus[idx1];
+            let two = &self.corpus[idx2];
+            let mut new_seq = vec![];
+            // TOOD rounds of mutations on elements?
+            match rng.gen_range(0..3) {
+                // TODO expose config and add tests
+                // splice
+                0 => {
+                    let start1 = rng.gen_range(0..one.len());
```

Comment:
there's an edge case here when generated `start1` is the same as `end1` and `start2` as `end2`. This will result in an empty `new_seq` which will then panic when performing runs in `initial_seq[0].clone(),`

```rust
    let mut current_run = InvariantTestRun::new(
        initial_seq[0].clone(),
        // Before each run, we must reset the backend state.
        self.executor.clone(),
        self.config.depth as usize,
);
```